File: sdk/javascript/verifai.js
/**
 * VerifAI JavaScript/Node.js SDK
 * Detect AI-generated videos with a single API call.
 *
 * Install:
 *   npm install node-fetch form-data
 *
 * Usage:
 *   const { VerifAI } = require('./verifai');
 *   const client = new VerifAI({ apiKey: 'your_key' });
 *   const result = await client.detectUrl('https://www.tiktok.com/@user/video/123');
 *   console.log(result.verdict); // "ai_generated" | "ai_edited" | "real"
 *
 * Get your API key: https://web-zeta-ecru-80.vercel.app/dashboard
 */

const API_BASE = 'https://ai-video-detector-production-a305.up.railway.app';

class DetectionResult {
  constructor(data, url = null) {
    this.verdict = data.verdict || 'unknown';
    this.isAiGenerated = data.is_ai_generated || false;
    this.isAiEdited = this.verdict === 'ai_edited';
    this.isReal = this.verdict === 'real';
    this.confidence = data.confidence || 0;
    this.confidencePct = data.confidence_pct || '0%';
    this.aiToolDetected = data.ai_tool_detected || null;
    this.editToolDetected = data.edit_tool_detected || null;
    this.detectionMethod = data.detection_method || '';
    this.url = url || data.url || null;
    this.filename = data.filename || null;
  }

  toString() {
    const labels = {
      ai_generated: '🤖 AI Generated',
      ai_edited: '✏️ AI Edited',
      real: '✅ Real',
    };
    const label = labels[this.verdict] || '❓ Unknown';
    const tool = this.aiToolDetected ? ` (${this.aiToolDetected})` : '';
    return `${label}${tool} — ${this.confidencePct} confidence`;
  }
}

class VerifAI {
  /**
   * @param {Object} options
   * @param {string} options.apiKey - Your VerifAI API key
   * @param {number} [options.timeout=60000] - Request timeout in ms
   */
  constructor({ apiKey, timeout = 60_000 } = {}) {
    if (!apiKey) throw new Error('VerifAI: apiKey is required');
    this.apiKey = apiKey;
    this.timeout = timeout;
    this._headers = {
      'X-Api-Key': apiKey,
      'Content-Type': 'application/json',
    };
  }

  async _fetch(path, options = {}) {
    const fetch = (await import('node-fetch')).default;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);
    try {
      const res = await fetch(`${API_BASE}${path}`, {
        ...options,
        signal: controller.signal,
      });
      if (!res.ok) {
        const err = await res.json().catch(() => ({}));
        throw new Error(`VerifAI API error ${res.status}: ${err.detail || res.statusText}`);
      }
      return res;
    } finally {
      clearTimeout(timer);
    }
  }

  /**
   * Detect if a video URL is AI-generated.
   * @param {string} url - Video URL (TikTok, Instagram, YouTube, etc.)
   * @param {boolean} [deep=false] - Run deeper visual analysis
   * @returns {Promise<DetectionResult>}
   */
  async detectUrl(url, deep = false) {
    const res = await this._fetch('/detect-url', {
      method: 'POST',
      headers: this._headers,
      body: JSON.stringify({ url, deep }),
    });
    const data = await res.json();
    return new DetectionResult(data, url);
  }

  /**
   * Detect if a video file is AI-generated.
   * @param {string|Buffer} fileOrPath - File path (Node.js) or Buffer
   * @param {string} [filename='video.mp4'] - Filename for the upload
   * @returns {Promise<DetectionResult>}
   */
  async detectFile(fileOrPath, filename = 'video.mp4') {
    const FormData = (await import('form-data')).default;
    const fs = await import('fs');
    const form = new FormData();
    const buffer = typeof fileOrPath === 'string'
      ? fs.readFileSync(fileOrPath)
      : fileOrPath;
    form.append('file', buffer, { filename, contentType: 'video/mp4' });
    const res = await this._fetch('/detect', {
      method: 'POST',
      headers: { 'X-Api-Key': this.apiKey, ...form.getHeaders() },
      body: form,
    });
    const data = await res.json();
    return new DetectionResult(data);
  }

  /**
   * Detect multiple URLs at once (Business/Enterprise tier).
   * @param {string[]} urls - Array of video URLs
   * @returns {Promise<DetectionResult[]>}
   */
  async detectBatch(urls) {
    const res = await this._fetch('/detect-batch', {
      method: 'POST',
      headers: this._headers,
      body: JSON.stringify(urls),
    });
    const items = await res.json();
    return items.map(item => new DetectionResult(item));
  }

  /**
   * Get your current API usage and limits.
   * @returns {Promise<Object>}
   */
  async usage() {
    const res = await this._fetch('/me', { headers: this._headers });
    return res.json();
  }
}

module.exports = { VerifAI, DetectionResult };

// ─── TypeScript types (paste into .d.ts file) ────────────────────────────────
/**
 * @typedef {'ai_generated' | 'ai_edited' | 'real' | 'unknown'} Verdict
 *
 * @typedef {Object} DetectionResultData
 * @property {Verdict} verdict
 * @property {boolean} isAiGenerated
 * @property {boolean} isAiEdited
 * @property {boolean} isReal
 * @property {number} confidence
 * @property {string} confidencePct
 * @property {string|null} aiToolDetected
 * @property {string|null} editToolDetected
 * @property {string} detectionMethod
 */

File: sdk/python/verifai.py
"""
VerifAI Python SDK
Detect AI-generated videos with a single API call.

Install:
    pip install requests

Usage:
    from verifai import VerifAI

    client = VerifAI(api_key="your_key")
    result = client.detect_url("https://www.tiktok.com/@user/video/123")
    print(result.verdict)   # "ai_generated" | "ai_edited" | "real"
    print(result.confidence) # 0.97
    print(result.ai_tool)    # "OpenAI Sora"

Get your API key: https://web-zeta-ecru-80.vercel.app/dashboard
"""
import requests
from dataclasses import dataclass
from typing import Optional, List
from pathlib import Path


API_BASE = "https://ai-video-detector-production-a305.up.railway.app"


@dataclass
class DetectionResult:
    verdict: str                    # "ai_generated" | "ai_edited" | "real" | "unknown"
    is_ai_generated: bool
    confidence: float               # 0.0 – 1.0
    confidence_pct: str             # "97.0%"
    ai_tool_detected: Optional[str] # "OpenAI Sora", "Kuaishou Kling", etc.
    edit_tool_detected: Optional[str]
    detection_method: str
    url: Optional[str] = None
    filename: Optional[str] = None
    # Audit-grade breakdown: deciding layer, per-layer scores, provenance
    # flags (C2PA, metadata stripped, platform re-encode) and caveats.
    explanation: Optional[dict] = None

    @property
    def is_real(self) -> bool:
        return self.verdict == "real"

    @property
    def is_ai_edited(self) -> bool:
        return self.verdict == "ai_edited"

    def __str__(self):
        label = {"ai_generated": "🤖 AI Generated", "ai_edited": "✏️ AI Edited", "real": "✅ Real"}.get(self.verdict, "❓ Unknown")
        tool = f" ({self.ai_tool_detected})" if self.ai_tool_detected else ""
        return f"{label}{tool} — {self.confidence_pct} confidence"


class VerifAI:
    """
    VerifAI client for AI video detection.

    Args:
        api_key: Your VerifAI API key. Get one at https://web-zeta-ecru-80.vercel.app/dashboard
        timeout: Request timeout in seconds (default: 60)
    """

    def __init__(self, api_key: str, timeout: int = 60):
        self.api_key = api_key
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({"X-Api-Key": api_key})

    def detect_url(self, url: str, deep: bool = False) -> DetectionResult:
        """
        Detect if a video URL is AI-generated.

        Args:
            url: Video URL (TikTok, Instagram, YouTube, etc.)
            deep: Run deeper visual analysis (slower, more accurate)

        Returns:
            DetectionResult with verdict and confidence

        Example:
            result = client.detect_url("https://www.tiktok.com/@user/video/123")
            if result.is_ai_generated:
                print(f"AI video! Tool: {result.ai_tool_detected}")
        """
        response = self.session.post(
            f"{API_BASE}/detect-url",
            json={"url": url, "deep": deep},
            timeout=self.timeout,
        )
        response.raise_for_status()
        data = response.json()
        return DetectionResult(
            verdict=data.get("verdict", "unknown"),
            is_ai_generated=data.get("is_ai_generated", False),
            confidence=data.get("confidence", 0.0),
            confidence_pct=data.get("confidence_pct", "0%"),
            ai_tool_detected=data.get("ai_tool_detected"),
            edit_tool_detected=data.get("edit_tool_detected"),
            detection_method=data.get("detection_method", ""),
            url=url,
            explanation=data.get("explanation"),
        )

    def detect_file(self, file_path: str) -> DetectionResult:
        """
        Detect if a video file is AI-generated.

        Args:
            file_path: Path to video file (.mp4, .mov, .webm, etc.)

        Returns:
            DetectionResult with verdict and confidence
        """
        path = Path(file_path)
        if not path.exists():
            raise FileNotFoundError(f"File not found: {file_path}")

        with open(path, "rb") as f:
            response = self.session.post(
                f"{API_BASE}/detect",
                files={"file": (path.name, f, "video/mp4")},
                timeout=self.timeout,
            )
        response.raise_for_status()
        data = response.json()
        return DetectionResult(
            verdict=data.get("verdict", "unknown"),
            is_ai_generated=data.get("is_ai_generated", False),
            confidence=data.get("confidence", 0.0),
            confidence_pct=data.get("confidence_pct", "0%"),
            ai_tool_detected=data.get("ai_tool_detected"),
            edit_tool_detected=data.get("edit_tool_detected"),
            detection_method=data.get("detection_method", ""),
            filename=path.name,
            explanation=data.get("explanation"),
        )

    def detect_batch(self, urls: List[str]) -> List[DetectionResult]:
        """
        Detect multiple URLs at once (Enterprise/Business tier required).

        Args:
            urls: List of video URLs to analyze

        Returns:
            List of DetectionResult, one per URL

        Note: Requires Business tier (100 URLs/batch) or Enterprise (1000 URLs/batch).
              Each URL counts as 1 request toward your monthly quota.
        """
        response = self.session.post(
            f"{API_BASE}/detect-batch",
            json=urls,
            timeout=self.timeout * len(urls),
            stream=True,
        )
        response.raise_for_status()
        import json
        results = []
        for item in json.loads(response.content):
            if "error" in item:
                results.append(DetectionResult(
                    verdict="unknown", is_ai_generated=False, confidence=0.0,
                    confidence_pct="0%", ai_tool_detected=None, edit_tool_detected=None,
                    detection_method=f"Error: {item['error']}", url=item.get("url"),
                ))
            else:
                results.append(DetectionResult(
                    verdict=item.get("verdict", "unknown"),
                    is_ai_generated=item.get("is_ai_generated", False),
                    confidence=item.get("confidence", 0.0),
                    confidence_pct=item.get("confidence_pct", "0%"),
                    ai_tool_detected=item.get("ai_tool_detected"),
                    edit_tool_detected=item.get("edit_tool_detected"),
                    detection_method=item.get("detection_method", ""),
                    url=item.get("url"),
                ))
        return results

    def usage(self) -> dict:
        """Returns your current API usage and limits."""
        response = self.session.get(f"{API_BASE}/me", timeout=10)
        response.raise_for_status()
        return response.json()


# ─── Quick-start example ──────────────────────────────────────────────────────

if __name__ == "__main__":
    import os
    key = os.getenv("VERIFAI_API_KEY", "your_api_key_here")
    client = VerifAI(api_key=key)

    # Single URL
    result = client.detect_url("https://www.youtube.com/shorts/3tmd-ClpJxA")
    print(f"Result: {result}")
    print(f"Verdict: {result.verdict}")
    print(f"Confidence: {result.confidence_pct}")

    # Check usage
    usage = client.usage()
    print(f"\nUsage: {usage.get('requests_this_month')}/{usage.get('monthly_limit')} this month")
